### PR TITLE
OHRM5X-2084: Develop - Claim request attachment API

### PR DESCRIPTION
**Pull request by DulsaraNethmin**

### CODE REVIEW

File: src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+            $claimAttachment = new ClaimAttachment();
+            $requestId = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, self::PARAMETER_REQUEST_ID);
+            $claimRequest = $this->getClaimService()->getClaimDao()->getClaimRequestById($requestId);
+            $this->throwRecordNotFoundExceptionIfNotExist($claimRequest, ClaimRequest::class);
+            if (!$this->getUserRoleManagerHelper()->isEmployeeAccessible($claimRequest->getEmployee()->getEmpNumber())) {
+                throw $this->getForbiddenException();
+            }
+            $claimAttachment->setRequestId($requestId);
+            $userId = $this->getAuthUser()->getUserId();
+            $claimAttachment->getDecorator()->setUserByUserId($userId);
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php:173]:
```suggestion
            $claimAttachment->getDecorator()->setUserByUserId($this->getAuthUser()->getUserId());
```

File: src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php [hunk 2]
```diff
@@ -315,4 +318,90 @@ public function deleteClaimExpense(ClaimRequest  $claimRequest, array $ids): int
...
+     */
+    public function getNextAttachmentId(int $requestId): int
+    {
+        $q = $this->createQueryBuilder(ClaimAttachment::class, 'attachment')
+            ->select('MAX(attachment.attachId)')
+            ->andWhere('attachment.requestId = :requestId')
+            ->setParameter('requestId', $requestId);
+        $id = $q->getQuery()->execute();
+        $id[0][1]++;
+        dump($id[0][1]);
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php:344]:
```suggestion
```

?????

**Review by ManulMax — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
+    ];
+
+    /**
+     * @OA\Get(
+     *     path="/api/v2/claim/requests/{requestId}/attachments",
+     *     tags={"Claim/Attachments"},
+     *     @OA\Parameter(
+     *         name="requestId",
+     *         in="attribute",
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php:79]: if `requestId` in the URL `in="attribute"` is correct

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php [hunk 4]
```diff
@@ -68,22 +68,20 @@ public function __construct(PartialClaimAttachment $claimAttachment)
...
+                'fileType',
+                'size',
+                'description',
                 'attachedTime'
             ]
         );
         $this->setAttributeNames(
             [
-                'requestId',
                 'attachmentId',
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php:81]:
```suggestion
                'id',
```

**Review by ManulMax — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseAPI.php [hunk 5]
```diff
@@ -172,27 +172,40 @@ public function setClaimExpense(ClaimExpense $claimExpense): void
...
+            $expenseTypeId = $this->getRequestParams()->getInt(
+                RequestParams::PARAM_TYPE_BODY,
+                self::PARAMETER_EXPENSE_TYPE_ID
+            );
+            $expenseType = $this->getClaimService()->getClaimDao()->getExpenseTypeById($expenseTypeId);
+            $this->throwRecordNotFoundExceptionIfNotExist($expenseType, ExpenseType::class);
+            $claimExpense->getDecorator()->setExpenseTypeByExpenseTypeId(
+                $this->getRequestParams()
+                    ->getInt(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_EXPENSE_TYPE_ID)
+            );
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseAPI.php:196]:
```suggestion
            $claimExpense->getDecorator()->setExpenseTypeByExpenseTypeId($expenseTypeId);
```

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+     *     @OA\Property(
+     *         property="fileDescription",
+     *         type="string"
+     *     ),
+     *     @OA\Property(
+     *         property="date",
+     *         type="string"
+     *     )
+     * )
+     */
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php:65]: move this doc comment on to class

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                'attachedTime'
+            ]
+        );
+        $this->setAttributeNames(
+            [
+                'id',
+                'fileName',
+                'fileType',
+                'size',
+                'fileDescription',
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php:85]:
```suggestion
                'description',
```

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                'date'
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php:86]: is this formatted, seems not used decorator

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimExpenseModel.php [hunk 9]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimExpenseModel.php:2]: remove this line

File: src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php [hunk 10]
```diff
@@ -296,14 +293,25 @@ public function getClaimExpenseById(int $id): ?ClaimExpense
...
+    {
+        return $this->getRepository(ClaimRequest::class)->findOneBy(['id' => $id, 'isDeleted' => false]);
     }
 
     /**
      * @param int[] $ids
      * @return int
      */
-    public function deleteClaimExpense(ClaimRequest  $claimRequest, array $ids): int
+    public function deleteClaimExpense(ClaimRequest $claimRequest, array $ids): int
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php:314]:
```suggestion
    public function deleteClaimExpense(int $claimRequestId, array $ids): int
```
shall we go like this :thinking: 
Also update function's doc comment

File: src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php [hunk 11]
```diff
@@ -315,4 +323,135 @@ public function deleteClaimExpense(ClaimRequest  $claimRequest, array $ids): int
...
+    }
+
+    /**
+     * @param int $requestId
+     * @return int
+     */
+    public function getNextAttachmentId(int $requestId): int
+    {
+        $q = $this->createQueryBuilder(ClaimAttachment::class, 'attachment')
+            ->select('MAX(attachment.attachId)')
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php:344]:
```suggestion
        $q = $this->createQueryBuilder(ClaimAttachment::class, 'attachment');
        $q->select($q->expr()->max('attachment.attachId')
```

File: src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php [hunk 12]
```diff
@@ -315,4 +323,135 @@ public function deleteClaimExpense(ClaimRequest  $claimRequest, array $ids): int
...
+            ->andWhere('attachment.requestId = :requestId')
+            ->setParameter('requestId', $requestId);
+        $id = $q->getQuery()->execute();
+        $id[0][1]++;
+        return $id[0][1];
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php:349]:
```suggestion
        $id = $q->getQuery()->getSingleScalarResult();
        $id = $id === null ? 0 : intval($id);
        return ++$id;
```

File: src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php [hunk 13]
```diff
@@ -315,4 +323,135 @@ public function deleteClaimExpense(ClaimRequest  $claimRequest, array $ids): int
...
+        $q = $this->createQueryBuilder(ClaimAttachment::class, 'claimAttachment')
+            ->select('count(claimAttachment.attachId)');
+        $this->setSortingAndPaginationParams($q, $claimAttachmentSearchFilterParams);
+        if (!is_null($claimAttachmentSearchFilterParams->getRequestId())) {
+            $requestId = $claimAttachmentSearchFilterParams->getRequestId();
+            $q->andWhere('claimAttachment.requestId = :requestId');
+            $q->setParameter('requestId', $requestId);
+        }
+        $this->setSortingAndPaginationParams($q, $claimAttachmentSearchFilterParams);
+        return $this->getPaginator($q)->getQuery()->execute()[0][1];
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php:428]:
```suggestion
        $q = $this->createQueryBuilder(ClaimAttachment::class, 'claimAttachment');
        $this->setSortingAndPaginationParams($q, $claimAttachmentSearchFilterParams);
        if (!is_null($claimAttachmentSearchFilterParams->getRequestId())) {
            $q->andWhere('claimAttachment.requestId = :requestId');
            $q->setParameter('requestId', $claimAttachmentSearchFilterParams->getRequestId());
        }
        return $this->getPaginator($q)->count();
```

File: src/plugins/orangehrmClaimPlugin/Dto/ClaimAttachmentSearchFilterParams.php [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Claim\Dto;
+
+use OrangeHRM\Core\Dto\FilterParams;
+
+class ClaimAttachmentSearchFilterParams extends FilterParams
+{
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Dto/ClaimAttachmentSearchFilterParams.php:25]: allowed sort fields not defined

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php [hunk 15]
```diff
@@ -0,0 +1,501 @@
...
+     *     @OA\RequestBody(
+     *         @OA\JsonContent(
+     *             type="object",
+     *             @OA\Property(property="attachment", type="object"
+     *                @OA\Property(property="name", type="string"),
+     *                @OA\Property(property="type", type="string"),
+     *                @OA\Property(property="size", type="string"),
+     *                @OA\Property(property="base64", type="base64"),
+     *             @OA\Property(property="description", type="string"),
+     *         )
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php:408]: use ref `#/components/schemas/Base64Attachment`

File: src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php [hunk 16]
```diff
@@ -0,0 +1,501 @@
...
+                CommonParams::PARAMETER_ID
+            );
+            $this->getClaimRequest($requestId);
+            $claimAttachment = $this->getClaimService()->getClaimDao()->getClaimAttachment($requestId, $attachId);
+            $this->throwRecordNotFoundExceptionIfNotExist($claimAttachment, ClaimAttachment::class);
+            $attachment = $this->getRequestParams()->getAttachmentOrNull(
+                RequestParams::PARAM_TYPE_BODY,
+                self::PARAMETER_CLAIM_ATTACHMENT_CONTENT
+            );
+            if ($attachment) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php:444]:
```suggestion
            if ($attachment instanceof Base64Attachment) {
```

[the remark this answers is not in the document]
Comment (RajithaKumara) [reply src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php:444]:
```suggestion
            if ($attachment instanceof Base64Attachment) {
```

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php (hunk 8 above)
Comment (RajithaKumara) [reply src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php:86]: check `src/plugins/orangehrmAttendancePlugin/Api/Model/AttendanceRecordModel.php` for employee object

**Review by ManulMax — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+ *         @OA\Property(property="empNumber", type="integer"),
+ *         @OA\Property(property="lastName", type="string"),
+ *         @OA\Property(property="firstName", type="string"),
+ *         @OA\Property(property="middleName", type="string"),
+ *         @OA\Property(property="employeeId", type="string"),
+ *     ),
+ *     @OA\Property(
+ *         property="date",
+ *         type="string"
+ *     )
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php:69]: @OA\Property(property="date", type="string", format="date"),

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+ *     @OA\Property(
+ *         property="fileSize",
+ *         type="integer"
+ *     ),
+ *     @OA\Property(
+ *         property="fileDescription",
+ *         type="string"
+ *     ),
+ *     @OA\Property(
+ *         property="Employee",
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimAttachmentModel.php:58]:
```suggestion
 *         property="employee",
```

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**